### ISSUE TITLE
[Bug] Joining Server fails

### ISSUE THREAD

**Issue by lionkor** [BeamMP/BeamMP-Server#171]
Since #165 was merged joining a server causes the following error:

```
[16/02/23 19:22:26] [ERROR] fatal: set_option: Invalid argument [system:22 at /usr/include/boos      t/asio/detail/reactive_socket_service.hpp:174 in function 'set_option']                              
[16/02/23 19:22:29] [ERROR] fatal: set_option: Invalid argument [system:22 at /usr/include/boos      t/asio/detail/reactive_socket_service.hpp:174 in function 'set_option']                              
[16/02/23 19:23:28] [ERROR] fatal: set_option: Invalid argument [system:22 at /usr/include/boos      t/asio/detail/reactive_socket_service.hpp:174 in function 'set_option'] 
```

**Comment by ghost** [comment]
Keep in mind that the alternative usage of the following does not appear to solve this problem:

```cpp
#include <boost/asio.hpp>
#include <boost/asio/ip/tcp.hpp>

using namespace boost::asio;
using namespace std::chrono;
```

I'm working with the code to see if I can find a better suited alternative or fix to this problem.

**Comment by ghost** [comment]
Just thought I'd provide an output log from my own shell. I am running on Ubuntu 22.04 LTS on a Microsoft Azure Virtual Machine, connected via SSH.

## Output Log from BeamMP-Server

```
azureuser@BeamMP-001:~/BeamMP-Server/build-linux$ sudo ./launch.bash
Terminating any running BeamMP-Server instances...
Starting BeamMP-Server...
Type 'exit' to close the server.
[28/03/23 19:37:09] TServer.cpp:81 [INFO] BeamMP Server v3.1.1
[28/03/23 19:37:09] TSentry.cpp:40 [INFO] Sentry started! Reporting errors automatically. This sends data to the developers in case of errors and crashes. You can learn more, turn this message off or opt-out of this in the ServerConfig.toml.
[28/03/23 19:37:09] TResourceManager.cpp:32 [INFO] Loaded 1 Mods
[28/03/23 19:37:09] TNetwork.cpp:80 [INFO] Vehicle data network online on port 30814 with a Max of 8 Clients
[28/03/23 19:37:09] TNetwork.cpp:144 [INFO] Vehicle event network online
[28/03/23 19:37:09] Common.cpp:212 [INFO] Server up-to-date!
[28/03/23 19:37:09] TPluginMonitor.cpp:30 [INFO] PluginMonitor started
[28/03/23 19:37:15] THeartbeatThread.cpp:115 [INFO] Authenticated! Success, Your server listing has been updated.
[28/03/23 19:37:15] main.cpp:196 [INFO] ALL SYSTEMS STARTED SUCCESSFULLY, EVERYTHING IS OKAY
> status
BeamMP-Server Status:
        Total Players:             0
        Syncing Players:           0
        Synced Players:            0
        Connected Players:         0
        Guests:                    0
        Cars:                      0
        Uptime:                    7882ms (~0h) 
        Lua:
                Queued results to check:     0
                States:                      1
                Event timers:                0
                Event handlers:              0
        Subsystems:
                Good/Starting/Bad:           12/0/0
                Shutting down/Shut down:     0/0
                Good:                        [ UpdateCheck, PluginMonitor, TCPNetwork, PPSMonitor, Sentry, UDPNetwork, Config, Heartbeat, ResourceManager, LuaEngine, Server, Main ]
                Starting:                    [  ]
                Bad:                         [  ]
                Shutting down:               [  ]
                Shut down:                   [  ]

[28/03/23 19:41:09] TNetwork.cpp:161 [ERROR] fatal: set_option: Invalid argument [system:22 at /usr/include/boost/asio/detail/reactive_socket_service.hpp:174 in function 'set_option']
> 
```

## Output Log from BeamMP-Launcher

```
[28/3/2023 12:40:41] [INFO] Launcher version is up to date
[28/3/2023 12:40:41] [INFO] Game Version : 0.27.2.0
[28/3/2023 12:40:41] [INFO] Downloading mod please wait...
Progress : [ 100% ] [##########################]
[28/3/2023 12:40:42] [INFO] Download Complete!
[28/3/2023 12:40:42] [INFO] Game Launched!
[28/3/2023 12:40:55] [INFO] Game Connected!
[28/3/2023 12:41:09] [INFO] Connecting to server
[28/3/2023 12:41:09] [INFO] Connected!
[28/3/2023 12:41:09] [INFO] Terminated!
[28/3/2023 12:41:09] [WARN] Failed to close socket!
```

**Comment by ghost** [comment]
One thought I am considering is to create a custom Boost Asio socket option to set the receive timeout on a Boost Asio socket.

```cpp
#include <boost/asio.hpp>

using boost::asio::ip::tcp;

struct rcv_timeout_option {
    rcv_timeout_option(int timeout) : timeout_(timeout) {}
    template <typename SettableSocketOption>
    void set_option(SettableSocketOption& option) {
        option.set_option(SO_RCVTIMEO, timeval{ timeout_ / 1000, (timeout_ % 1000) * 1000 });
    }
private:
    int timeout_;
};
```

This is merely an example and I am testing to see if this is a viable option.

**Comment by ghost** [comment]
## Changes

Keep in mind this new method requires a new struct definition for the custom option to be used as an argument to the set_option() function.

```cpp
#include <boost/asio.hpp>

using boost::asio::ip::tcp;

struct rcv_timeout_custom_option {
    rcv_timeout_custom_option(int timeout) : timeout_(timeout) {}
    template <typename Protocol>
    void set_option(tcp::socket& socket, boost::system::error_code& ec) const {
        boost::asio::detail::socket_option::integer<IPPROTO_TCP, TCP_KEEPIDLE> option(timeout_ / 1000);
        socket.set_option(option, ec);
    }
private:
    int timeout_;
};
```

## Usage

The usage should be done as such:

```cpp
try {
    if (Application::IsShuttingDown()) {
        beammp_debug("shutdown during TCP wait for accept loop");
        break;
    }
    ip::tcp::endpoint ClientEp;
    ip::tcp::socket ClientSocket = Acceptor.accept(ClientEp, ec);
    if (ec) {
        beammp_errorf("failed to accept: {}", ec.message());
    }
    rcv_timeout_custom_option option(120000);
    TConnection Conn { std::move(ClientSocket), ClientEp };
    std::thread ID(&TNetwork::Identify, this, std::move(Conn));
    ID.detach(); // TODO: Add to a queue and attempt to join periodically
} catch (const std::exception& e) {
    beammp_error("fatal: " + std::string(e.what()));
}
```

## Testing

**THESE CHANGES ARE NOT TESTED**

Simply put, this is just a suggestion. I am running tests at the moment.

**Comment by ghost** [comment]
The changes have been tested and have worked! This issue can be closed and you may update your code to the corresponding fix above.

I am glad I was able to help :)

## Launcher Log

```
[28/3/2023 12:40:41] [INFO] Launcher version is up to date
[28/3/2023 12:40:41] [INFO] Game Version : 0.27.2.0
[28/3/2023 12:40:41] [INFO] Downloading mod please wait...
Progress : [ 100% ] [##########################]
[28/3/2023 12:40:42] [INFO] Download Complete!
[28/3/2023 12:40:42] [INFO] Game Launched!
[28/3/2023 12:40:55] [INFO] Game Connected!
[28/3/2023 12:41:09] [INFO] Connecting to server
[28/3/2023 12:41:09] [INFO] Connected!
[28/3/2023 12:41:09] [INFO] Terminated!
[28/3/2023 12:41:09] [WARN] Failed to close socket!
[28/3/2023 13:18:37] [WARN] Failed to close socket!
[28/3/2023 13:18:39] [INFO] Connecting to server
[28/3/2023 13:18:40] [INFO] Connected!
[28/3/2023 13:18:40] [INFO] Checking Resources...
[28/3/2023 13:18:40] [INFO] Syncing...
[28/3/2023 13:18:41] [INFO] Done!
```

## Shell & Server Logs

```
azureuser@BeamMP-001:~/BeamMP-Server/build-linux$ sudo make install
[  2%] Built target fmt
[ 24%] Built target breakpad_client
[ 57%] Built target sentry
[ 59%] Built target commandline
[ 60%] Building CXX object CMakeFiles/BeamMP-Server.dir/src/TNetwork.cpp.o
/home/azureuser/BeamMP-Server/src/TNetwork.cpp: In member function ‘std::vector<unsigned char> TNetwork::TCPRcv(TClient&)’:
/home/azureuser/BeamMP-Server/src/TNetwork.cpp:462:11: warning: comparison of integer expressions of different signedness: ‘long unsigned int’ and ‘int32_t’ {aka ‘int’} [-Wsign-compare]
  462 |     if (N != Header) {
      |         ~~^~~~~~~~~
             [ 60%] Building CXX object CMakeFiles/BeamMP-Server.dir/src/LuaAPI.cpp.o
[ 61%] Building CXX object CMakeFiles/BeamMP-Server.dir/src/TPluginMonitor.cpp.o
[ 62%] Linking CXX executable BeamMP-Server
[ 78%] Built target BeamMP-Server
[ 79%] Building CXX object CMakeFiles/BeamMP-Server-tests.dir/src/TConsole.cpp.o
[ 79%] Building CXX object CMakeFiles/BeamMP-Server-tests.dir/src/TServer.cpp.o
/home/azureuser/BeamMP-Server/src/TServer.cpp: In static member function ‘static bool TServer::IsUnicycle(TClient&, const string&)’:
/home/azureuser/BeamMP-Server/src/TServer.cpp:238:35: warning: unused parameter ‘c’ [-Wunused-parameter]
  238 | bool TServer::IsUnicycle(TClient& c, const std::string& CarJson) {
      |                          ~~~~~~~~~^
[ 80%] Building CXX object CMakeFiles/BeamMP-Server-tests.dir/src/TLuaEngine.cpp.o
[ 81%] Building CXX object CMakeFiles/BeamMP-Server-tests.dir/src/TLuaPlugin.cpp.o
[ 82%] Building CXX object CMakeFiles/BeamMP-Server-tests.dir/src/Http.cpp.o
/home/azureuser/BeamMP-Server/src/Http.cpp: In lambda function:
/home/azureuser/BeamMP-Server/src/Http.cpp:195:66: warning: unused parameter ‘Req’ [-Wunused-parameter]
  195 |     HttpLibServerInstance->set_logger([](const httplib::Request& Req, const httplib::Response& Res) {
      |                                          ~~~~~~~~~~~~~~~~~~~~~~~~^~~
/home/azureuser/BeamMP-Server/src/Http.cpp:195:96: warning: unused parameter ‘Res’ [-Wunused-parameter]
  195 |     HttpLibServerInstance->set_logger([](const httplib::Request& Req, const httplib::Response& Res) {
      |                                                                       ~~~~~~~~~~~~~~~~~~~~~~~~~^~~
[ 83%] Building CXX object CMakeFiles/BeamMP-Server-tests.dir/src/TPPSMonitor.cpp.o
[ 84%] Building CXX object CMakeFiles/BeamMP-Server-tests.dir/src/TNetwork.cpp.o
/home/azureuser/BeamMP-Server/src/TNetwork.cpp: In member function ‘std::vector<unsigned char> TNetwork::TCPRcv(TClient&)’:
/home/azureuser/BeamMP-Server/src/TNetwork.cpp:462:11: warning: comparison of integer expressions of different signedness: ‘long unsigned int’ and ‘int32_t’ {aka ‘int’} [-Wsign-compare]
  462 |     if (N != Header) {
      |         ~~^~~~~~~~~
[ 85%] Building CXX object CMakeFiles/BeamMP-Server-tests.dir/src/LuaAPI.cpp.o
/home/azureuser/BeamMP-Server/src/LuaAPI.cpp: In function ‘int LuaAPI::PanicHandler(lua_State*)’:
/home/azureuser/BeamMP-Server/src/LuaAPI.cpp:304:37: warning: unused parameter ‘State’ [-Wunused-parameter]
  304 | int LuaAPI::PanicHandler(lua_State* State) {
      |                          ~~~~~~~~~~~^~~~~
[ 86%] Building CXX object CMakeFiles/BeamMP-Server-tests.dir/src/TPluginMonitor.cpp.o
[ 87%] Linking CXX executable BeamMP-Server-tests
[ 97%] Built target BeamMP-Server-tests
[100%] Built target test
Install the project...
-- Install configuration: "Release"
-- Up-to-date: /usr/local/lib/cmake/sol2/sol2-targets.cmake
-- Up-to-date: /usr/local/include/sol
-- Up-to-date: /usr/local/include/sol/detail
-- Up-to-date: /usr/local/include/sol/detail/build_version.hpp
-- Up-to-date: /usr/local/include/sol/traits.hpp
-- Up-to-date: /usr/local/include/sol/function.hpp
-- Up-to-date: /usr/local/include/sol/stack_check_get_unqualified.hpp
-- Up-to-date: /usr/local/include/sol/optional.hpp
-- Up-to-date: /usr/local/include/sol/bind_traits.hpp
-- Up-to-date: /usr/local/include/sol/sol.hpp
-- Up-to-date: /usr/local/include/sol/lua_table.hpp
-- Up-to-date: /usr/local/include/sol/stack_core.hpp
-- Up-to-date: /usr/local/include/sol/function_types_templated.hpp
-- Up-to-date: /usr/local/include/sol/usertype_proxy.hpp
-- Up-to-date: /usr/local/include/sol/thread.hpp
-- Up-to-date: /usr/local/include/sol/as_args.hpp
-- Up-to-date: /usr/local/include/sol/object.hpp
-- Up-to-date: /usr/local/include/sol/stack_check.hpp
-- Up-to-date: /usr/local/include/sol/as_returns.hpp
-- Up-to-date: /usr/local/include/sol/demangle.hpp
-- Up-to-date: /usr/local/include/sol/usertype_container.hpp
-- Up-to-date: /usr/local/include/sol/variadic_results.hpp
-- Up-to-date: /usr/local/include/sol/stack_probe.hpp
-- Up-to-date: /usr/local/include/sol/trampoline.hpp
-- Up-to-date: /usr/local/include/sol/bytecode.hpp
-- Up-to-date: /usr/local/include/sol/function_result.hpp
-- Up-to-date: /usr/local/include/sol/stack_get.hpp
-- Up-to-date: /usr/local/include/sol/make_reference.hpp
-- Up-to-date: /usr/local/include/sol/ebco.hpp
-- Up-to-date: /usr/local/include/sol/table_iterator.hpp
-- Up-to-date: /usr/local/include/sol/prologue.hpp
-- Up-to-date: /usr/local/include/sol/wrapper.hpp
-- Up-to-date: /usr/local/include/sol/base_traits.hpp
-- Up-to-date: /usr/local/include/sol/debug.hpp
-- Up-to-date: /usr/local/include/sol/proxy_base.hpp
-- Up-to-date: /usr/local/include/sol/pairs_iterator.hpp
-- Up-to-date: /usr/local/include/sol/error.hpp
-- Up-to-date: /usr/local/include/sol/pointer_like.hpp
-- Up-to-date: /usr/local/include/sol/forward_detail.hpp
-- Up-to-date: /usr/local/include/sol/deprecate.hpp
-- Up-to-date: /usr/local/include/sol/stack_iterator.hpp
-- Up-to-date: /usr/local/include/sol/version.hpp
-- Up-to-date: /usr/local/include/sol/table_proxy.hpp
-- Up-to-date: /usr/local/include/sol/stack_check_get_qualified.hpp
-- Up-to-date: /usr/local/include/sol/protect.hpp
-- Up-to-date: /usr/local/include/sol/stack_check_get.hpp
-- Up-to-date: /usr/local/include/sol/function_types_core.hpp
-- Up-to-date: /usr/local/include/sol/state.hpp
-- Up-to-date: /usr/local/include/sol/table.hpp
-- Up-to-date: /usr/local/include/sol/stack_push.hpp
-- Up-to-date: /usr/local/include/sol/usertype_storage.hpp
-- Up-to-date: /usr/local/include/sol/stack.hpp
-- Up-to-date: /usr/local/include/sol/optional_implementation.hpp
-- Up-to-date: /usr/local/include/sol/usertype.hpp
-- Up-to-date: /usr/local/include/sol/compatibility
-- Up-to-date: /usr/local/include/sol/compatibility/compat-5.4.h
-- Up-to-date: /usr/local/include/sol/compatibility/lua_version.hpp
-- Up-to-date: /usr/local/include/sol/compatibility/compat-5.3.c.h
-- Up-to-date: /usr/local/include/sol/compatibility/compat-5.3.h
-- Up-to-date: /usr/local/include/sol/policies.hpp
-- Up-to-date: /usr/local/include/sol/coroutine.hpp
-- Up-to-date: /usr/local/include/sol/call.hpp
-- Up-to-date: /usr/local/include/sol/in_place.hpp
-- Up-to-date: /usr/local/include/sol/stack_proxy_base.hpp
-- Up-to-date: /usr/local/include/sol/stack_get_qualified.hpp
-- Up-to-date: /usr/local/include/sol/resolve.hpp
-- Up-to-date: /usr/local/include/sol/stack_check_unqualified.hpp
-- Up-to-date: /usr/local/include/sol/function_types.hpp
-- Up-to-date: /usr/local/include/sol/raii.hpp
-- Up-to-date: /usr/local/include/sol/lua_value.hpp
-- Up-to-date: /usr/local/include/sol/overload.hpp
-- Up-to-date: /usr/local/include/sol/stack_guard.hpp
-- Up-to-date: /usr/local/include/sol/protected_handler.hpp
-- Up-to-date: /usr/local/include/sol/inheritance.hpp
-- Up-to-date: /usr/local/include/sol/types.hpp
-- Up-to-date: /usr/local/include/sol/config.hpp
-- Up-to-date: /usr/local/include/sol/stack_get_unqualified.hpp
-- Up-to-date: /usr/local/include/sol/reference.hpp
-- Up-to-date: /usr/local/include/sol/assert.hpp
-- Up-to-date: /usr/local/include/sol/property.hpp
-- Up-to-date: /usr/local/include/sol/usertype_core.hpp
-- Up-to-date: /usr/local/include/sol/object_base.hpp
-- Up-to-date: /usr/local/include/sol/protected_function_result.hpp
-- Up-to-date: /usr/local/include/sol/load_result.hpp
-- Up-to-date: /usr/local/include/sol/stack
-- Up-to-date: /usr/local/include/sol/stack/detail
-- Up-to-date: /usr/local/include/sol/stack/detail/pairs.hpp
-- Up-to-date: /usr/local/include/sol/string_view.hpp
-- Up-to-date: /usr/local/include/sol/protected_function.hpp
-- Up-to-date: /usr/local/include/sol/function_types_stateful.hpp
-- Up-to-date: /usr/local/include/sol/unique_usertype_traits.hpp
-- Up-to-date: /usr/local/include/sol/unsafe_function.hpp
-- Up-to-date: /usr/local/include/sol/error_handler.hpp
-- Up-to-date: /usr/local/include/sol/epilogue.hpp
-- Up-to-date: /usr/local/include/sol/state_handling.hpp
-- Up-to-date: /usr/local/include/sol/state_view.hpp
-- Up-to-date: /usr/local/include/sol/function_types_overloaded.hpp
-- Up-to-date: /usr/local/include/sol/compatibility.hpp
-- Up-to-date: /usr/local/include/sol/tie.hpp
-- Up-to-date: /usr/local/include/sol/userdata.hpp
-- Up-to-date: /usr/local/include/sol/packaged_coroutine.hpp
-- Up-to-date: /usr/local/include/sol/table_core.hpp
-- Up-to-date: /usr/local/include/sol/variadic_args.hpp
-- Up-to-date: /usr/local/include/sol/function_types_stateless.hpp
-- Up-to-date: /usr/local/include/sol/stack_pop.hpp
-- Up-to-date: /usr/local/include/sol/metatable.hpp
-- Up-to-date: /usr/local/include/sol/dump_handler.hpp
-- Up-to-date: /usr/local/include/sol/stack_proxy.hpp
-- Up-to-date: /usr/local/include/sol/environment.hpp
-- Up-to-date: /usr/local/include/sol/stack_check_qualified.hpp
-- Up-to-date: /usr/local/include/sol/tuple.hpp
-- Up-to-date: /usr/local/include/sol/usertype_container_launch.hpp
-- Up-to-date: /usr/local/include/sol/unsafe_function_result.hpp
-- Up-to-date: /usr/local/include/sol/unicode.hpp
-- Up-to-date: /usr/local/include/sol/stack_reference.hpp
-- Up-to-date: /usr/local/include/sol/stack_field.hpp
-- Up-to-date: /usr/local/include/sol/usertype_traits.hpp
-- Up-to-date: /usr/local/include/sol/forward.hpp
-- Up-to-date: /usr/local/lib/cmake/sol2/sol2-config.cmake
-- Up-to-date: /usr/local/lib/cmake/sol2/sol2-config-version.cmake
-- Up-to-date: /usr/local/share/pkgconfig/sol2.pc
-- Up-to-date: /usr/local/include/doctest/doctest.h
-- Up-to-date: /usr/local/lib/cmake/doctest/doctestConfig.cmake
-- Up-to-date: /usr/local/lib/cmake/doctest/doctestConfigVersion.cmake
-- Up-to-date: /usr/local/lib/cmake/doctest/doctest.cmake
-- Up-to-date: /usr/local/lib/cmake/doctest/doctestAddTests.cmake
-- Up-to-date: /usr/local/lib/cmake/doctest/doctestTargets.cmake
azureuser@BeamMP-001:~/BeamMP-Server/build-linux$ cd ~/BeamMP-Server/build-linux
azureuser@BeamMP-001:~/BeamMP-Server/build-linux$ sudo ./launch.bash
Terminating any running BeamMP-Server instances...
Starting BeamMP-Server...
Type 'exit' to close the server.
[28/03/23 20:18:31] TServer.cpp:81 [INFO] BeamMP Server v3.1.1
[28/03/23 20:18:31] TSentry.cpp:40 [INFO] Sentry started! Reporting errors automatically. This sends data to the developers in case of errors and crashes. You can learn more, turn this message off or opt-out of this in the ServerConfig.toml.
[28/03/23 20:18:31] TResourceManager.cpp:32 [INFO] Loaded 1 Mods
[28/03/23 20:18:31] TNetwork.cpp:158 [INFO] Vehicle event network online
[28/03/23 20:18:31] TNetwork.cpp:94 [INFO] Vehicle data network online on port 30814 with a Max of 8 Clients
[28/03/23 20:18:31] Common.cpp:212 [INFO] Server up-to-date!
[28/03/23 20:18:31] TPluginMonitor.cpp:30 [INFO] PluginMonitor started
[28/03/23 20:18:36] THeartbeatThread.cpp:115 [INFO] Authenticated! Success, Your server listing has been updated.
[28/03/23 20:18:36] main.cpp:196 [INFO] ALL SYSTEMS STARTED SUCCESSFULLY, EVERYTHING IS OKAY
[28/03/23 20:18:40] TNetwork.cpp:246 [INFO] Identifying new ClientConnection...
[28/03/23 20:18:41] TNetwork.cpp:368 [INFO] Identification success
[28/03/23 20:18:41] TNetwork.cpp:626 [INFO] Client connected
[28/03/23 20:18:41] TNetwork.cpp:629 [INFO] Assigned ID 0 to shawnjb
[28/03/23 20:18:41] TNetwork.cpp:683 [INFO] shawnjb requesting : /replication.zip
[28/03/23 20:18:41] TNetwork.cpp:635 [INFO] shawnjb : Connected
[28/03/23 20:19:05] TNetwork.cpp:908 [INFO] shawnjb is now synced!
> 
```

**Comment by lionkor** [comment]
Hi, please make a fork, branch and PR with these changes, otherwise it's very difficult to follow or test them. Thank you! Mention this issue in the commit title & PR title, like so: `fix xyz (closes #171)`.

**Comment by lionkor** [comment]
Oh, you did, sorry, it's early ;)
